### ISSUE TITLE
[Enhancement] choco-url to easily install a package via one click like with AptProtocol (apt://)

### ISSUE THREAD

**Issue by DJCrashdummy** [chocolatey/choco#1680]
i'm not sure where to post this, so please bear with me if this is the wrong place!

---

i like [AptProtocol](https://wiki.debian.org/AptProtocol) for its easiness to suggest installation of a package either on a website, in an email or in any other document via one click (resp. a second one to confirm it's installation).

IMHO a choco-protocol - or maybe also work with `apt://` - would be a nice tool to provide the simplest method possible to install packages from chocolatey.
e.g. `chocolateygui` is nice but i rather use https://chocolatey.org/packages to browse the packages because it provides more information. after that i have open powershell and type the install-command (which is ok), but noobs have to open `chocolateygui`, wait for a refresh and look for the correct package again to install it. instead of this bunch of additional steps and time a little "install now"-button at the homepage would ease the things a lot.

**Comment by gep13** [comment]
@AdmiringWorm @DJCrashdummy if I have understood this correctly, we would need the ability to pass a navigation parameter into Chocolatey GUI to directly navigate to a Package Details page.  Or have I misunderstood this?

**Comment by gep13** [comment]
@AdmiringWorm which registry keys are you referring to?  Do you have a link to what you are describing?

**Comment by bcurran3** [comment]
Could this be accomplished as well via a browser plugin to open nuspkg files using "choco install" upon download?

**Comment by AdmiringWorm** [comment]
> @bcurran3 said:
> Could this be accomplished as well via a browser plugin to open nuspkg files using "choco install" upon download?

I honestly don't know, but it probably could.

Anyways, I tested out the registry changes, and it does indeed work. Unfortunately due to the nature of chocolatey where it needs to run as an admin, this is something I haven't yet been able to do 😢

**Comment by bcurran3** [comment]
Setting cinst.exe as "Run this program as an administrator"

and

> Windows Registry Editor Version 5.00
> 
> [HKEY_CLASSES_ROOT\nupkg_auto_file]
> 
> [HKEY_CLASSES_ROOT\nupkg_auto_file\shell]
> 
> [HKEY_CLASSES_ROOT\nupkg_auto_file\shell\open]
> 
> [HKEY_CLASSES_ROOT\nupkg_auto_file\shell\open\command]
> @="\"C:\\ProgramData\\chocolatey\\bin\\cinst.exe\" \"%1\""

...allows double clicking on .nupkg files to autorun/autoinstall

Maybe there's something to run with here...

**Comment by DJCrashdummy** [comment]
> > if I have understood this correctly, we would need the ability to pass a navigation parameter into Chocolatey GUI to directly navigate to a Package Details page. Or have I misunderstood this?
>
> No, to the CLI itself. And not to navigate, but to actually install (unless I misunderstood it myself).

@gep13 and @AdmiringWorm yes and no:
for sure the execution is done at `choco.exe` but i highly recommend to somehow ask for confirmation before doing so, because hardly any noob looks at the URL behind a link before he clicks it. - doing this in the shell will scare off noobs, so some kind of GUI should be involved.

at debian/ubuntu/etc. `apturl` is a separate package and it doesn't open any package manager, but a small window to ask for a confirmation:
![apturl](https://user-images.githubusercontent.com/26110463/48846569-bbbb5700-ed9f-11e8-8f64-2ce083190605.png)

an other not so advanced approach would be to open the existing package details page of `chocolateygui`... but still an improvement compared to now. - and what the URL does/opens can still be changed/improved later.

**Comment by AdmiringWorm** [comment]
@DJCrashdummy I don't know about other browsers, but I do know that chrome asks for confirmation to run the executable associated with the custom protocol.

IMO, this should be enough (if other browsers behave the same)

**Comment by DJCrashdummy** [comment]
maybe it is useful... a wiki-link of `apturl` with integration steps into some browsers:
https://help.ubuntu.com//community/AptURL

**Comment by bcurran3** [comment]
In the spirit of the original request...
And dedicated to @DJCrashdummy.

I present to you:
https://chocolatey.org/packages/choco-install-packages-from-web-winconfig

**Comment by bcurran3** [comment]
@DJCrashdummy - your original dream is becoming a reality...

https://chocolatey.org/packages/chocoprotocolstub

**Comment by bcurran3** [comment]
The following is my personal re-write of the enhancement request:

> Chocolatey - like yum or apt-get, but for Windows

Requesting choco.exe feature duplication of AptURL/Apt Protocol. 
e.g. chocoURL/choco Protocol

Implementation:
* Upon installation of Chocolatey a custom choco Protocol would be configured in the Window's registry.
* choco.exe input to accept chocoURL (choco://packagename) as an input OR creation of a choco protocol handler/stub (SUGGEST: cprot.exe)
* All package pages updated on chocolatey.org :-) and new corporate intranet web pages with click to install corporate approved software.

Considerations:
* When browsers run the program associated with the protocol, the configured program will run with typically non-elevated user permissions. This results in only a handful of packages that could actually be installed. Thus, there needs be some way to elevate to admin privileges for real world package installation. To solve this, choco.exe could be passed a new argument that tells it to re-run itself with elevated privileges. Using a pre-processing choco protocol handler/stub with the internal manifest set to run as admin might be a better way of doing this. Going the choco protocol handler/stub route, no modification to choco.exe would be necessary. The choco protocol would be configured to run the choco protocol handler/stub. The chocoURL would be parsed by the choco protocol handler/stub (running with admin privs) and would then run choco.exe to install the package.

Proof of concept implemented using a handler/stub program I wrote to process the chocoURL and then execute cinst.exe. https://chocolatey.org/packages/chocoprotocolstub/1.0.0.0

**Comment by DJCrashdummy** [comment]
beforehand: sorry for being silent so long, but i was busy at work the last months and had no time to fiddle around with a windows machine.

---

### `choco-protocol-support`
thank you very much @bcurran3! :+1: it works like a charm (even with edge - i've tested it because it is nowhere mentioned). :clap:
**but** one interesting thing i noticed: (whether it is installed or not) if i try to view the details-page in the *Chocolatey GUI* it crushes. - this is the only package with this issue so i guess it must have to do with it and/or its description resp. rendering of them in the *Chocolatey GUI*.

### `choco-install-packages-from-web-winconfig`
unfortunately it does not work for me. :slightly_frowning_face: 
i'm pretty sure because `cinst.exe` is **not** executed as admin (although it is explained otherwise in the details). :confused: 

PS @bcurran3: may i suggest to strike through all mentions and links of <s>`chocoprotocolstub`</s> since it was rejected... or is there a reason why it may still be interesting to try?

**Comment by bcurran3** [comment]
> thank you very much @bcurran3! 👍 it works like a charm

![](https://i.pinimg.com/736x/e5/29/4e/e5294e9fffbafa212eff2e40211f359e.jpg)

> (even with edge - i've tested it because it is nowhere mentioned). 👏

I did some QUICK testing with edge and it looked like it sandboxed it so quit really fast and moved on. :) Thanks for letting us (me) know it works with Edge!

> **but** one interesting thing i noticed: (whether it is installed or not) if i try to view the details-page in the _Chocolatey GUI_ it crushes. - this is the only package with this issue so i guess it must have to do with it and/or its description resp. rendering of them in the _Chocolatey GUI_.

Very weird. Not sure what's causing that. I thought maybe the choco protocol URL was crashing it but it displays it fine in other packages. I opened https://github.com/chocolatey/ChocolateyGUI/issues/649 for @gep13 to check it out.

> ### `choco-install-packages-from-web-winconfig`
unfortunately it does not work for me. 🙁
i'm pretty sure because `cinst.exe` is **not** executed as admin (although it is explained otherwise in the details). 😕

I'd suggest to remove and reinstall. There's also a possibility that it might not work if you already had .nupkg associated with a program under your USER profile. BUT... I'm putting out a v0.0.2 soon reworked to run a batchfile and use Sudo  to keep it inline with it's sister package choco-protocol-support. That way it doesn't need to change the cinst.exe compatibility layer/manifest

> PS @bcurran3: may i suggest to strike through all mentions and links of `chocoprotocolstub` since it was rejected... or is there a reason why it may still be interesting to try?

Done. I didn't at first because it works, and works slightly better and faster than choco-protocol-stub and people can still use it if they wanted. It was rejected not because it didn't work or work well, @ferventcoder didn't "want chocolatey.org to be the first validation in the "web of trust" for assemblies that are placed onto chocolatey.org. Rather, that this validation is asserted elsewhere.
What it comes down to is that although we know you as a maintainer and trust the work that you have done in this area, we can't independently verify what is going on with the assembly that you are adding into the package." So basically it wasn't allowed because they didn't have the source. So I re-imagined it into a batch file that anyone could see what it does. I have gone back and striked it out and edited in a comment that it was replaced/superceded by [choco-protocol-support](https://chocolatey.org/packages/choco-protocol-support) which does the same thing and is "blessed."

**Comment by DJCrashdummy** [comment]
> (I don't use Edge.) Thanks for letting us (me) know it works with Edge!

me neither, :wink: but i didn't have something else on a virgin test machine. - that was my intention.

> I thought maybe the choco protocol URL was crashing it but it displays it fine in other packages.

this was also my first thought, but it is definitely not the case. - yes, very weird.

> choco-install-packages-from-web-winconfig
I'd suggest to remove and reinstall. [...] I'm putting out a v0.0.2 soon reworked [...]

already tried this and now once again, but no change. i just wanted to let you know... - (since i hardly use windows) no problem, i'll "wait" for v0.0.2.

**Comment by Omzig** [comment]
There is another option that might be easier: 

Check out the install button in this link:
 https://marketplace.visualstudio.com/items?itemName=auchenberg.vscode-browser-preview

It is: vscode:extension/auchenberg.vscode-browser-preview
so, the format is:
choco:action/package

This is the documentation for it:
https://docs.microsoft.com/en-us/windows/uwp/launch-resume/launch-app-with-uri

It is the same idea as the **mailto:** link.

**Comment by DJCrashdummy** [comment]
although not suggested at my initial post... if not using `apt://` but "creating" an own `choco://` protocol, supporting chocolaty parameters as discussed at https://github.com/bcurran3/ChocolateyPackages/issues/138 would make sense.

**Comment by TheCakeIsNaOH** [comment]
After thinking about this some and some discussion with @gep13, there are a couple of points about how this should be implemented:
- Instead of passing the URI to `choco install`, a new command like `choco uri` should be created. This is for security reasons, so someone can't do something like `choco install choco://googlechrome --yes --source=https://malware/api`. Having a separate command should allow much more secure parsing.
- It would be good to have a versioned api, so any future updates/changes can be made, and we are not locked into one format.  E.g. start off with `choco://v1/<package to install>`, then later add more features like `choco:///v2/<install|upgrade|uninstall>/<package>/<version>?parameter=param` or whatever.
- If specifying a source is allowed should be thought through carefully.

**Comment by TheCakeIsNaOH** [comment]
This actually appears to be a duplicate of #684, although most of the discussion is happening in this issue.

**Comment by gep13** [comment]
This issue was added to the 1.0.0 milestone, however, after discussion, we are not going to include this in the 1.0.0 release, but hopefully quite quickly after the 1.0.0 release.

**Comment by Omzig** [comment]
can we get this moved out of the backlog and into the next to be worked on pile?

**Comment by pauby** [comment]
If you have a look at the [top issues in the repository](https://github.com/chocolatey/choco/issues/3708), this is not ranked highly in priority for users.

**Comment by Omzig** [comment]
If i understand this correctly, we need to make the 4 thumbs up at the top go over 18, so it will surface in that list :)

**Comment by Omzig** [comment]
> i'm not sure where to post this, so please bear with me if this is the wrong place!
> 
> i like [AptProtocol](https://wiki.debian.org/AptProtocol) for its easiness to suggest installation of a package either on a website, in an email or in any other document via one click (resp. a second one to confirm it's installation).
> 
> IMHO a choco-protocol - or maybe also work with `apt://` - would be a nice tool to provide the simplest method possible to install packages from chocolatey. e.g. `chocolateygui` is nice but i rather use https://chocolatey.org/packages to browse the packages because it provides more information. after that i have open powershell and type the install-command (which is ok), but noobs have to open `chocolateygui`, wait for a refresh and look for the correct package again to install it. instead of this bunch of additional steps and time a little "install now"-button at the homepage would ease the things a lot.

this the first post please, we need it to go over 18

**Comment by Omzig** [comment]
woohoo we are up to 14 now, only a few more now ;)

**Comment by fdcastel** [comment]
@pauby 

> If you have a look at the [top issues in the repository](https://github.com/chocolatey/choco/issues/3708), this is not ranked highly in priority for users.

I just realized that ALL the ["top issues"](https://github.com/chocolatey/choco/issues/3708) date back to before 2019.

This isn't particularly inspiring, to put it mildly.

**Comment by fdcastel** [comment]
@gep13 

> This issue was added to the 1.0.0 milestone, however, after discussion, we are not going to include this in the 1.0.0 release, but hopefully quite quickly after the 1.0.0 release.

Do you recall what kept it out of the 1.0.0 release? If so, is that still a blocker? What would take to move this forward now?

@bcurran3’s implementation looks great. And I’d be happy to contribute if the team’s up for it.

**Comment by Omzig** [comment]
@fdcastel, yep, i have a bad feeling that no one is maintaining this much anymore.

**Comment by gep13** [comment]
> @Omzig said...
yep, i have a bad feeling that no one is maintaining this much anymore.

That is quite simply not true!

Chocolatey CLI has had 5 releases in the last 12 months.  You can see the release notes here:

https://docs.chocolatey.org/en-us/choco/release-notes/

Do we have a backlog of issues that people would _like_ to see added to Chocolatey CLI?  Yes, that is abundantly clear, but to say that no one is maintaing this project anymore is quite simply _rude_.

Do the priorities of the Chocolatey Development Team match your own priorities, such as the addition of a choco-url? No, probably not, however again, that doesn't mean that no work is being done.

The recently introduced Top Issues Dashboard is an attempt to allow popular bugs/feature requests to be shown to us, due to the fact that there are lots of issues in the backlog, and it can be hard to see through them all.

Is this a perfect solution to the problem?  Again, no, but it is something that we will be monitoring going forward, and it will help to establish what the priorities are.  Does it mean that every bug/feature request on the Top Issues Dashboard _will_ be tackled?  No, that is not the intention of the dashboard, it is a tool to use for helping with priorities, nothing more.

**Comment by gep13** [comment]
> @fdcastel said...
Do you recall what kept it out of the 1.0.0 release? If so, is that still a blocker? What would take to move this forward now?

Off the top of my head, I am not sure.  I believe I have some notes on this topic somewhere though, so I would need to go through them and refresh my memory.  I personally love this feature idea, and it is one that I would like to see added to Chocolatey CLI.  I do remember there being some complexities on _how_ this would work, and that was part of the reason that it stalled, but again, I would need to look back on my notes, and partial implementation work that was done at the time.

> @fdcastel said...
 And I’d be happy to contribute if the team’s up for it.

This is definitely something that we would welcome, but I would ask for you to hold off until a clearer direction on how to move forward with this feature is established.  I wouldn't want you to do a bunch of work, and then for us not to be able to merge it in.

Also, as you will no doubt have seen, we have a number of [open PR's](https://github.com/chocolatey/choco/pulls), some of which have been around for a while, so if you were to create a PR, we don't offer any guarantee on how long it will take for that PR to be merged.  Ask @TheCakeIsNaOH if you want some ideas on how long things can take.   In the last release of Chocolatey CLI, the number of open PR's went from over 50, down to 19.  I like to say that around here the "wheels turn slowly, but they do turn".  We are here, and we are listening, it just sometimes takes longer for things to happen than we would like.

**Comment by Omzig** [comment]
@gep13, my apologies. Was not my intent. We have however made it to 23 thumbs up, maybe that is enough to get it looked at ;)

**Comment by gep13** [comment]
> @Omzig said...
my apologies. Was not my intent. We have however made it to 23 thumbs up, maybe that is enough to get it looked at ;)

It certainly _helps_ to bring attention to the feature, yes, however, please don't mistake the Top Issues Dashboard as a guarantee that things are going to get implemented, that is _not_ what it is intended for.  It is a tool to help _guide_ the priorities of the team, not a guarantee for things to get implemented.

**Comment by Omzig** [comment]
Once this is implemented, then they can put a button on the choco website like the vscode marketplace and just click 'install'.

**Comment by Omzig** [comment]
@gep13, btw, for me it doesn't make much sense for this to be in the CLI, it makes more sense for the feature to be installed with the Chocolatey GUI, the GUI installer can put in the necessary registry entries where as the CLI install is a little more problematic.

**Comment by DJCrashdummy** [comment]
@Omzig IMHO this doesn't make much sense in the general Chocolatey GUI... as it is hardly needed at all in the mentioned workflow.
only a little window to ask for confirmation should be shown to be on the safe side (for those who click links without checking them). ...perhaps subsequently also a little progress bar for impatient people resp. big programs to know when the installation is finished.
**but** even if these get a graphical implementation and not just the usual CLI, i would prefer to keep them separate that you don't need to install the GUI.

thinking about it a bit further... what might also work and don't need additional packages: implement everything necessary in the core and make the choco-URL work with/in the CLI... and as soon as the GUI is installed use a GUI for confirmation & progress.

**Comment by Omzig** [comment]
@DJCrashdummy, oo maybe just a `Chocolatey Agent` install to insert a few lines of registry and call it done.

**Comment by fdcastel** [comment]
@gep13 

> (...) but I would ask for you to hold off until a clearer direction on how to move forward with this feature is established. I wouldn't want you to do a bunch of work, and then for us not to be able to merge it in.

Sure! I will never submit a PR without first discussing it with the team. As I see it, this would only add unnecessary burden for the maintainers rather than help move things forward.

Thanks again to you and the whole team for the amazing work over the years. And don’t hesitate to let me know if and when there’s something I can help with!
